analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/ParameterReassignedConstraintTest.cs

```diff
@@ -0,0 +1,90 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+using SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp;
+using SonarAnalyzer.UnitTest.TestFramework.SymbolicExecution;
+
+namespace SonarAnalyzer.UnitTest.Rules;
+
+[TestClass]
+public class ParameterReassignedConstraintTest
+{
+    [DataTestMethod]
+    [DataRow(@"arg = Unknown<object>();", "object")]
+    [DataRow(@"arg = Unknown<List<object>>();", "List<object>")]
+    [DataRow(@"arg = Unknown<int>();", "int")]
+    [DataRow(@"arg = Unknown<int?>();", "int?")]
+    [DataRow(@"arg = Unknown<object>();", "ref object")]
+    [DataRow(@"arg = Unknown<object>();", "out object")]
+    [DataRow(@"arg = Unknown<object[]>();", "params object[]")]
+    [DataRow(@"arg = Unknown<object>(); arg.ToString();", "object")]
+    [DataRow(@"arg = Unknown<object>(); arg.ToString();", "object")]
+    [DataRow(@"arg = Unknown<object>(); arg = Unknown<object>();", "object")]
+    public void ParameterReassignedConstraint_AfterAssignment(string methodSnippet, string argumentType)
+    {
+        var methodBody = $@"{methodSnippet} Tag(""AfterAssignment"", arg);";
+        var argumentSnippet = $", {argumentType} arg";
+        var validator = SETestContext.CreateCS(methodBody, argumentSnippet, new PublicMethodArgumentsShouldBeCheckedForNull()).Validator;
+        validator.ValidateTag("AfterAssignment", x => x.HasConstraint<ParameterReassignedConstraint>().Should().BeTrue());
```

Comment:
Use HaveOnlyConstraint() and the like. This applies to all the following tests as well (The `HasConstraint` based tests are the old way to do the assertion. We don't use it anymore for new tests and there will be an effort to replace all the old tests with the new way at one point).
```suggestion
        validator.ValidateTag("AfterAssignment", x => x.Should().HaveOnlyConstraint(ParameterReassignedConstraint.Instance));
```

Reply:
In this case, I only want to check whether the symbol has the `ParameterReassignedConstraint`. I don't mind if it has other constraints. e.g. 

`arg = Unknown<object>(); arg.ToString();`

In this case `arg` will have both the `ParameterReassigned` and `NotNull` constraints.

Using `HaveOnlyConstraint` would also make the test cases more fragile. If a new constraint is introduced (e.g. `InstanceMethodInvoked` or something similar) then it will break the tests.

Reply:
> Using HaveOnlyConstraint would also make the test cases more fragile. If a new constraint is introduced (e.g. InstanceMethodInvoked or something similar) then it will break the tests.

Yes. That is why we want to use HaveOnlyConstraint(s). We want such changes to be visible in all affected cases.

Reply:
You may need to split your test into two or more.